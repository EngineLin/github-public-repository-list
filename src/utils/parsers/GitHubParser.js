class GitHubParser {
  // Have to had those properties by the guild line.
  title;
  description;
  url;

  // Other properties.
  id;
  createdAt;
  gitUrl;

  constructor(
    name = "",
    description = "",
    html_url = "",
    id = -1,
    create_at = "",
    git_url = ""
  ) {
    this.title = name;
    this.description = description;
    this.url = html_url;
    this.id = id;
    this.createdAt = create_at;
    this.gitUrl = git_url;
  }

  static parse({ name, description, html_url, id, create_at, git_url } = {}) {
    return new GitHubParser(
      name,
      description,
      html_url,
      id,
      create_at,
      git_url
    );
  }

  static empty() {
    return new GitHubParser();
  }
}

export default GitHubParser;
