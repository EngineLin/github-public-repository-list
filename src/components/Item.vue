<template>
  <div class="container">
    <h3>{{ item.title }}</h3>

    <div class="wrapper">
      <p class="label">Description:</p>
      <p :class="{ warning: !item.description }">
        {{ item.description || "Has not description in this project." }}
      </p>
    </div>

    <div class="wrapper">
      <p class="label">Link:</p>
      <a :href="item.url" target="_blank" rel="noopener noreferrer nofollow">
        <p>{{ item.url }}</p>
      </a>
    </div>
  </div>
</template>

<script>
import GitHubParser from "@/utils/parsers/GitHubParser";

export default {
  name: "Item",
  props: {
    item: {
      type: GitHubParser,
      default: () => GitHubParser.empty()
    }
  }
};
</script>

<style scoped>
p {
  margin: 0;
}

.container {
  width: 100%;
  padding: 20px 0;
  margin: 20px 0;
  border-radius: 10px;
  box-sizing: border-box;
  background: white;
  box-shadow: 4px 4px 12px 1px rgba(0, 0, 0, 0.2);
  line-height: 1.5;
}

h3 {
  font-family: "Fira Mono", monospace;
  color: #783f8e;
  margin: 0 0 20px 0;
  letter-spacing: 1px;
}

.label {
  color: #438a5e;
  font-size: 18px;
  font-weight: 600;
  margin-right: 15px;
}

.wrapper {
  padding: 10px 20px;
  display: flex;
  align-items: center;
}

.warning {
  color: #fbd46d;
}

a {
  underline: unset;
}
</style>
