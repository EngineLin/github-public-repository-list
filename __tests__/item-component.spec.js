import "@testing-library/jest-dom";
import { render } from "@testing-library/vue";

import Item from "@/components/Item.vue";
import GitHubParser from "@/utils/parsers/GitHubParser";

test("Item component should render correct data with GitHubParser", () => {
  const tData = {
    name: "TEST_NAME",
    description: "TEST_DESCRIPTION",
    html_url: "https://test-url.com",
    id: 0,
    created_at: "2018-07-13T09:23:53Z"
  };
  const item = GitHubParser.parse(tData);
  const { getByText } = render(Item, { propsData: { item } });
  getByText(item.title);
  getByText(item.description);
  getByText(item.url);
  getByText(item.createdAt);
});

test("Item component should render with GitHubParser.empty() when props item has no data", () => {
  const { baseElement } = render(Item);
  expect(baseElement).toMatchSnapshot();
});
