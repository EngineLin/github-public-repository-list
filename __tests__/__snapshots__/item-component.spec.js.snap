// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item component should render with GitHubParser.empty() when props item has no data 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <h3>
        unknown
      </h3>
       
      <div
        class="wrapper"
      >
        <p
          class="label"
        >
          Description:
        </p>
         
        <p
          class="warning"
        >
          
      Has not description in this project.
    
        </p>
      </div>
       
      <div
        class="wrapper"
      >
        <p
          class="label"
        >
          Link:
        </p>
         
        <a
          href="#"
          rel="noopener noreferrer nofollow"
          target="_blank"
        >
          <p>
            unknown
          </p>
        </a>
      </div>
       
      <p
        class="created-at"
      >
        <span
          style="color: red;"
        >
          Created at 
        </span>
        
    unknown
  
      </p>
    </div>
  </div>
</body>
`;
